modules/javafx.graphics/src/main/native-glass/gtk/glass_window_ime.cpp

```diff
@@ -23,266 +23,157 @@
-            break;
-        }
-    }
+    nativePos = mainEnv->GetDoubleArrayElements(pos, NULL);
 
-    XFree(styles);
+    GdkRectangle rect;
+    if (nativePos) {
+        rect.x = (int) nativePos[0];
+        rect.y = (int) nativePos[1];
+        rect.width = 0;
+        rect.height = 0;
 
-    return result;
+        mainEnv->ReleaseDoubleArrayElements(pos, nativePos, 0);
+        gtk_im_context_set_cursor_location(im_ctx.ctx, &rect);
+    }
 }
 
 void WindowContextBase::enableOrResetIME() {
-    Display *display = gdk_x11_display_get_xdisplay(gdk_window_get_display(gdk_window));
-    if (xim.im == NULL || xim.ic == NULL) {
-        xim.im = XOpenIM(display, NULL, NULL, NULL);
-        if (xim.im == NULL) {
-            return;
-        }
-
-        XIMStyle styles = get_best_supported_style(xim.im);
-        if (styles == 0) {
-            return;
-        }
-
-        XIMCallback startCallback = {(XPointer) jview, (XIMProc) (void *) im_preedit_start};
-        XIMCallback doneCallback = {(XPointer) jview, im_preedit_done};
-        XIMCallback drawCallback = {(XPointer) jview, im_preedit_draw};
-        XIMCallback caretCallback = {(XPointer) jview, im_preedit_caret};
-
-        XVaNestedList list = XVaCreateNestedList(0,
-                XNPreeditStartCallback, &startCallback,
-                XNPreeditDoneCallback, &doneCallback,
-                XNPreeditDrawCallback, &drawCallback,
-                XNPreeditCaretCallback, &caretCallback,
-                NULL);
-
-        xim.ic = XCreateIC(xim.im,
-                XNInputStyle, styles,
-                XNClientWindow, GDK_WINDOW_XID(gdk_window),
-                XNPreeditAttributes, list,
-                NULL);
-
-        XFree(list);
-
-        if (xim.ic == NULL) {
-            return;
-        }
+    if (im_ctx.on_preedit) {
+        gtk_im_context_focus_out(im_ctx.ctx);
     }
 
-    if (xim.enabled) { //called when changed focus to different input
-        XmbResetIC(xim.ic);
+    if (!im_ctx.enabled) {
+        im_ctx.ctx = gtk_im_multicontext_new();
+        gtk_im_context_set_client_window(GTK_IM_CONTEXT(im_ctx.ctx), gdk_window);
+        g_signal_connect(im_ctx.ctx, "preedit-start", G_CALLBACK(on_preedit_start), this);
+        g_signal_connect(im_ctx.ctx, "preedit-changed", G_CALLBACK(on_preedit_changed), this);
+        g_signal_connect(im_ctx.ctx, "preedit-end", G_CALLBACK(on_preedit_end), this);
+        g_signal_connect(im_ctx.ctx, "commit", G_CALLBACK(on_commit), this);
     }
 
+    gtk_im_context_reset(im_ctx.ctx);
+    gtk_im_context_focus_in(im_ctx.ctx);
 
-    XSetICFocus(xim.ic);
-
-    xim.enabled = TRUE;
+    im_ctx.on_preedit = false;
+    im_ctx.enabled = true;
 }
 
 void WindowContextBase::disableIME() {
-    if (xim.ic != NULL) {
-        XUnsetICFocus(xim.ic);
+    if (im_ctx.ctx != NULL) {
+        g_signal_handlers_disconnect_matched(im_ctx.ctx, G_SIGNAL_MATCH_DATA, 0, 0, NULL, NULL, NULL);
+        g_object_unref(im_ctx.ctx);
```

Comment:
If the IM window is visible when the window is closed disableIME() can get called twice; I'm seeing debug output being generated. Set im_ctx.ctx to NULL here or you'll unref it twice.

Reply:
Fixed.

Reply:
Looks good to me.